[AERIE-1725]  get command typescript types

dyst5422 opened the pull request:
* **Tickets addressed:** AERIE-1725
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description

This PR will retrieve the command typescript Library used to author the command expansions.  Here is the workflow:

1. Use a Hasura action and provide the `commandDictionaryID`. 
2. The commanding server will retrieve the command typescript from the filesystem. Note: this was generated when the command dictionary was uploaded 
4. The library is base64 encoded and sent back as the response to the Hasura action

## Verification
I tested this locally

1. `docker system prune --all --volumes` nuke everything 
2. `./gradlew clean; ./gradlew assemble; docker compose -f docker-compose.yml build; docker compose -f docker-compose.yml up` 
3. Follow the step in this PR to upload the command dictionary. Start at step 3: https://github.com/NASA-AMMOS/aerie/pull/70
5. Run the action below. 

```
query commandTypescript {
  getCommandTypeScript(commandDictionaryId: "1") {
    typescript
  }
}
```
6. Verify that you receive `base64` string back from the action.
7. Take the string and convert it with this URL: https://www.base64decode.org/


## Documentation
I will have to add these steps to the User Guide that we will be giving the users

## Future work
Creating Hasura action for the other endpoints in the commanding server and hooking up the rest of the commanding server.

camargo commented on docker-compose.yml:
```diff
@@ -92,7 +95,6 @@ services:
       GATEWAY_SERVER_URL: http://aerie_gateway:9000
       HASURA_CLIENT_URL: http://localhost:8080/v1/graphql
       HASURA_SERVER_URL: http://hasura:8080/v1/graphql
-      NODE_TLS_REJECT_UNAUTHORIZED: "0"
```
We should keep this since this is the development docker-compose file

goetzrrGit replied:
@camargo this was fixed when I rebased and updated this PR. We can probably resolve this

goetzrrGit committed:
Hook up the hasura action to the app.post("/get-command-typescript") endpoint

This will retrieve the command typescript library file as a base64 string. The user will have to convert it on their end.

[AERIE-1725]

goetzrrGit committed:
Add getCommandTypescriptType class

This method will query the database to find the command typescript library path and read it from the file system. This library was created when the user uploads the command dictionary

goetzrrGit committed:
Added a new Hasura Actions to get the commandTypescript

 Input:

     `query command {
       getCommandTypeScript(commandDictionaryId: "1") {
         typescript
       }
     }`

 Output:

   `"data": {
     "getCommandTypeScript": {
       "typescript": base64 String
       }
     }`

goetzrrGit commented on deployment/hasura/metadata/actions.graphql:
```diff
@@ -25,6 +25,19 @@ type Query {
   ): EffectiveArgumentsResponse
 }
 
+type Query {
+  getActivityTypeScript(
+    missionModelId: ID!
+    activityTypeName: String!
+  ): TypeScriptResponse
+}
```
I cherry picked getActivityTypeScript from https://github.com/NASA-AMMOS/aerie/pull/81

mattdailis commented on command-expansion-server/src/app.ts:
```diff
@@ -70,8 +71,15 @@ app.put("/expansion-set", async (req, res) => {
+app.post("/command-typescript", async (req, res) => {
+  const commandDictionaryId: string = req.body.input.commandDictionaryId;
+  const commandTypescript = await getCommandTypescriptTypes(db, parseInt(commandDictionaryId, 10));
+  const commandTypescriptBase64 = Buffer.from(commandTypescript).toString("base64");
```
Not going to block this PR on this - but worth revisiting later whether we prefer base64 to JSON.stringify here

dyst5422 replied:
I prefer the base64 option because its more simple than json.stringify (not changing actual bytestream) and is probably more widely available than JSON.stringify

goetzrrGit replied:
I agree with Dylan. base64 is easier to work with. So far Upload, activity typescript generation, and command typescript library are using base64 when being transferred through the Hasura actions.

mattdailis reviewed (approved):
Looks good to me. There's some overlap with the activity typescript PR (namely the hasura metadata), but that can be sorted out when rebasing the other onto develop.

dyst5422 commented:
@goetzrrGit Do we need these to all be POSTs or can we make them specific methods? (not sure whats available in the actions). I ask because we have the expansion get/put and we'd have to either change our route naming or change our methods to better organize that (with the method being preferred I think?)

goetzrrGit commented:
@dyst5422 all Hasura query/mutation are POST so we should change the URL to 

`app.post("/get-command-typescript"`

I can do that now and rebase.

dyst5422 commented:
> @dyst5422 all Hasura query/mutation are POST so we should change the URL to
> 
> `app.post("/get-command-typescript"`
> 
> I can do that now and rebase.

Yeah, I can submit a PR to adjust the ones that are already merged to follow that convention <method>-<name>

goetzrrGit commented:
@dyst5422 yeah just make it a hotfix so we can approve it faster. Everyone seems to review hotfixes really quick

goetzrrGit commented:
Updated the endpoint and hasura to `app.post('get-command-typescript')`

goetzrrGit commented on deployment/hasura/metadata/actions.yaml:
```diff
@@ -1,79 +1,83 @@
+    definition:
+      kind: ""
+      handler: http://aerie_commanding:3000/get-activity-typescript
+  - name: getCommandTypeScript
+    definition:
+      kind: ""
+      handler: http://aerie_commanding:3000/get-command-typescript
```
`getCommandTypeScript` was added for this PR
